src/site/content/en/blog/porting-gphoto2-to-the-web/index.md

```diff
@@ -0,0 +1,354 @@
+
+```js
+while (this.canvasRef.current) {
+  try {
+    let blob = await this.props.getPreview();
+
+    let img = await createImageBitmap(blob, { /* … */ });
+    await new Promise(resolve => requestAnimationFrame(resolve));
+    canvasCtx.transferFromImageBitmap(img);
+  } catch (err) {
+    // …
+  }
+}
+```
+
+Using those modern APIs ensures that all of the decoding work is done in the background, and the canvas gets updated only when both the image and the browser are fully prepared for drawing. This achieved a consistent 30+ FPS on my laptop, which matched the native performance of both gPhoto2 and the official Sony software.
+
+## Synchronizing the USB access
+
+When a USB data transfer is requested while another operation is already in progress, it will commonly result in a "device is busy" error. Since the preview and the settings UI update regularly, and the user might be trying to capture an image or modify settings at the same time, such conflicts between different operations turned out to be very frequent.
+
+To avoid them, I needed to synchronize all the accesses within the application. For that, I've built a promise-based async queue:
+
+```js
+let context = await new Module.Context();
+
+let queue = Promise.resolve();
+
+function schedule(op) {
+  let res = queue.then(() => op(context));
+  queue = res.catch(rethrowIfCritical);
+  return res;
+}
+```
+
+By chaining each operation in a `then()` callback of the existing `queue` promise, and storing the chained result as the new value of `queue`, I can make sure that all operations are executed one by one, in order and without overlaps.
+
+Any operation errors are returned to the caller, while critical (unexpected) errors mark the entire chain as a rejected promise, and ensure that no new operation will be scheduled afterwards.
+
+By keeping the module context in a private (non-exported) variable, I'm minimizing the risks of accessing the `context` by accident somewhere else in the app without going through the `schedule()` call.
+
+To tie things together, now each access to the device context has to be wrapped in a `schedule()` call like this:
+
+```js
+let config = await this.connection.schedule((context) => context.configToJS());
+```
+
+and
+
+```js
+this.connection.schedule((context) => context.captureImageAsFile());
+```
+
+After that, all the operations were executing successfully without conflicts.
+
+## Conclusion
+
+Feel free to browse the [codebase on Github](https://github.com/GoogleChromeLabs/web-gphoto2) for more implementation insights. I also want to thank [Marcus Meissner](https://github.com/msmeissn) for maintenance of gPhoto2 and for his reviews of my upstream PRs.
```

Comment:
@jpmedley note: added one small paragraph here because I was pointed out I don't have a direct link to the repo but it might be useful for other readers.